pkg/agent/controller/ipseccertificate/ipsec_certificate_controller.go

```diff
@@ -0,0 +1,520 @@
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(obj)
+	if key, ok := obj.(string); !ok {
+		c.queue.Forget(obj)
+		klog.Errorf("Expected string in work queue but got %#v", obj)
+		return true
+	} else if err := c.syncConfigurations(); err == nil {
+		c.queue.Forget(key)
+	} else {
+		c.queue.AddRateLimited(key)
+		klog.Errorf("Error syncing IPSec certificates, requeuing. Error: %v", err)
+	}
+	return true
+}
+
+type certificateKeyPair struct {
+	// rawCert is an optional field to determine if signing cert/key pairs have changed
+	rawCert []byte
+	// rawKey is an optional field to determine if signing cert/key pairs have changed
+	rawKey []byte
+	// rawCert is an optional field to determine if root CA have changed
+	rawCACert []byte
+
+	caCertificate []*x509.Certificate
+	certificate   []*x509.Certificate
+	privateKey    crypto.Signer
+}
+
+func (pair *certificateKeyPair) equals(target *certificateKeyPair) bool {
+	return bytes.Equal(pair.rawCACert, target.rawCACert) &&
+		bytes.Equal(pair.rawKey, target.rawKey) &&
+		bytes.Equal(pair.rawCert, target.rawCert)
+}
+
+func (pair *certificateKeyPair) validate() error {
+	if len(pair.caCertificate) == 0 {
+		return fmt.Errorf("CA certificate is empty")
+	}
+	if len(pair.certificate) == 0 {
+		return fmt.Errorf("certificate is empty")
+	}
+	if pair.privateKey == nil {
+		return fmt.Errorf("private key is empty")
+	}
+	roots := x509.NewCertPool()
+	for _, r := range pair.caCertificate {
+		roots.AddCert(r)
+	}
+	certificate := pair.certificate[0]
+	verifyOptions := x509.VerifyOptions{
+		Roots: roots,
+		KeyUsages: []x509.ExtKeyUsage{
+			x509.ExtKeyUsageIPSECTunnel,
+		},
+	}
+	if _, err := certificate.Verify(verifyOptions); err != nil {
+		return err
+	}
+	switch pub := certificate.PublicKey.(type) {
+	case *rsa.PublicKey:
+		priv, ok := pair.privateKey.(*rsa.PrivateKey)
+		if !ok {
+			return fmt.Errorf("private key type does not match public key type")
+		}
+		if pub.N.Cmp(priv.N) != 0 {
+			return fmt.Errorf("private key does not match public key")
+		}
+	default:
+		return fmt.Errorf("unrecognized certificate public key type")
+	}
+	return nil
+}
+
+var jitteryDuration = func(totalDuration float64) time.Duration {
+	return wait.Jitter(time.Duration(totalDuration), 0.2) - time.Duration(totalDuration*0.3)
+}
+
+// nextRotationDeadline returns a value for the threshold at which the
+// current certificate should be rotated, 80%+/-10% of the expiration of the
+// certificate.
+func (pair *certificateKeyPair) nextRotationDeadline() time.Time {
+	if err := pair.validate(); err != nil {
+		klog.ErrorS(err, "Verify certificate configurations failed")
+		return time.Now()
+	}
+	notAfter := pair.certificate[0].NotAfter
+	totalDuration := float64(notAfter.Sub(pair.certificate[0].NotBefore))
+	deadline := pair.certificate[0].NotBefore.Add(jitteryDuration(totalDuration))
```

Comment:
I think it cann't count deadline starting from `NotBefore`, otherwise the certificate may expire in 1 day from now but `NotAfter - NotBefore` can be 1 year.

Reply:
the deadline is an exact time calculated based on the certificate's total duration and some jitter.
we compare the time with now to determine whether or not to start rotation.

Reply:
I see, thanks for explanation